astropy/io/ascii/sextractor.py

```diff
@@ -36,14 +36,18 @@ def get_cols(self, lines):
         # However, some may be missing and must be inferred from skipped column numbers
         columns = {}
         # E.g. '# 1 ID identification number' (without units) or '# 2 MAGERR magnitude of error [mag]'
-        re_name_def = re.compile(r"""^ \s* \# \s*            # possible whitespace around #
+        # Updated along with issue #4603, for more robust parsing of unit
+        re_name_def = re.compile(r"""^\s* \# \s*             # possible whitespace around #
                                  (?P<colnumber> [0-9]+)\s+   # number of the column in table
                                  (?P<colname> [-\w]+)        # name of the column
-                                 (?:\s+(?P<coldescr> \w [^\[]*\w))? # column description, match non-[
-                                 (?:\s+\[(?P<colunit>.+)\])?.*   # match units in brackets
+                                 (?:\s+(?P<coldescr> \w .+)  # column description, match any character until...
+                                 (?:(?<!(\]))$|(?=(?:(?<=\S)\s+\[.+\]))))?  # ...until [non-space][space][unit] or [not-right-bracket][end]
```

Comment:
The reason was that I observed the previous regex fails on parsing the description and unit for the last line of my sextractor3.dat file:

    #  14 MAGERR_APER            RMS error vector for fixed aperture mag.                   [mag]

Due to the fact that there is a trailing `.` at the description part, the previous regex will match `<coldescr>` until before `.`, and leave the rest (starting with the `.`) to `<colunit>`, which fails it, as the unit part assumes to be starting with `[`.
Similar thing would happen for this column as well:

    #   3 ALPHA_J2000            Right ascension of barycenter (J2000)                      [deg] 

where the previous regex returns description `Right ascension of barycenter (J2000`, and unit `None`.